metadata-ingestion/src/datahub/ingestion/source/metabase.py

```diff
@@ -565,10 +579,19 @@ def get_datasource_from_id(self, datasource_id):
                 reason=f"Platform was not found in DataHub. Using {platform} name as is",
             )
 
+        # For cases when metabase has several platform instances (e.g. several individual ClickHouse clusters)
+        datasource_id_in_metabase = dataset_json.get("id")
+        platform_instance = (
+            self.config.database_id_to_instance_map.get(str(datasource_id_in_metabase))
+            if datasource_id_in_metabase and self.config.database_id_to_instance_map
+            else None
+        )
+
+        # If Metabase datasource ID is not mapped to platform instace, fall back to platform mapping
         # Set platform_instance if configuration provides a mapping from platform name to instance
         platform_instance = (
             self.config.platform_instance_map.get(platform)
-            if self.config.platform_instance_map
+            if self.config.platform_instance_map and platform_instance is None
             else None
```

Comment:
This if needs to be written as below for this to behave correctly for case platform_instance was chosen from `datasource_id_in_metabase` .

```
if self.config.platform_instance_map and platform_instance is None:
     platform_instance = self.config.platform_instance_map.get(platform)
```

Can you please refractor the entire get platform instance logic in separate method and add unit tests for that method ?

```
def get_platform_instance_for_datasource(datasource_id: str, platform: Optional[str])->Optional[str]:
    ...
```

Reply:
I have moved platform_instance detection into method.
Attempted to add a unit-test for this function but due to connection checks being done by `__init__()` of `MetabaseSource` I need to also mock http communication. Will continue working on this.

Reply:
@mayurinehate would you please check the PR once again? I've implemented the unit-test for the function and it passes normally.

I've also moved network initialization of `MetabaseSource` class into separate function which is called in `MetabaseSource.__init__()`. There are no changes in behavior but now init is logically split into "local" initialization and "network" initialization.

I see some tests are failing but these are not related to the change I made.